fdbserver/RestoreMaster.actor.cpp

```diff
@@ -0,0 +1,1339 @@
+
+	loop {
+		try {
+			wait(lockDatabase(cx, uid));
+			break;
+		} catch( Error &e ) {
+			printf("Transaction Error when we lockDB. Error:%s\n", e.what());
+			wait(tr->onError(e));
+		}
+	}
+
+	state Reference<ReadYourWritesTransaction> tr(new ReadYourWritesTransaction(cx));
+	loop {
+		try {
+			tr->reset();
+			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+
+			wait(checkDatabaseLock(tr, uid));
+
+			tr->commit();
+			break;
+		} catch( Error &e ) {
+			printf("Transaction Error when we lockDB. Error:%s\n", e.what());
+			wait(tr->onError(e));
+		}
+	}
+
+
+	return Void();
+}
+
+ACTOR static Future<Void> _clearDB(Reference<ReadYourWritesTransaction> tr) {
+	loop {
+		try {
+			tr->reset();
+			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+			tr->clear(normalKeys);
+			tr->commit();
+			break;
+		} catch(Error &e) {
+			printf("Retry at clean up DB before restore. error code:%d message:%s. Retry...\n", e.code(), e.what());
+			if(e.code() != error_code_restore_duplicate_tag) {
+				wait(tr->onError(e));
+			}
+		}
+	}
+
+	return Void();
+}
+
+
+
+ACTOR Future<Void> initializeVersionBatch(Reference<RestoreMasterData> self) {
+	loop {
+		try {
+			wait(delay(1.0));
+			std::vector<Future<RestoreCommonReply>> cmdReplies;
+			self->cmdID.initPhase(RestoreCommandEnum::Reset_VersionBatch);
+			for (auto &loader : self->loadersInterf) {
+				cmdReplies.push_back( loader.second.initVersionBatch.getReply(RestoreVersionBatchRequest(self->cmdID, self->batchIndex)) );
+			}
+			for (auto &applier : self->appliersInterf) {
+				cmdReplies.push_back( applier.second.initVersionBatch.getReply(RestoreVersionBatchRequest(self->cmdID, self->batchIndex)) );
+			}
+
+			std::vector<RestoreCommonReply> reps = wait( timeoutError(getAll(cmdReplies), FastRestore_Failure_Timeout) );
+			printf("Initilaize Version Batch done\n");
+			break;
+		} catch (Error &e) {
+			fprintf(stdout, "[ERROR] Node:%s, Current phase: initializeVersionBatch, Commands before cmdID:%s error. error code:%d, error message:%s\n", self->describeNode().c_str(),
+					self->cmdID.toString().c_str(), e.code(), e.what());
+		}
+	}
+
+	return Void();
+}
+
+
+ACTOR Future<Void> notifyApplierToApplyMutations(Reference<RestoreMasterData> self) {
+	state std::vector<Future<RestoreCommonReply>> cmdReplies;
+	loop {
+		try {
+			self->cmdID.initPhase( RestoreCommandEnum::Apply_Mutation_To_DB );
+			state std::map<UID, RestoreApplierInterface>::iterator applier;
+			for (applier = self->appliersInterf.begin(); applier != self->appliersInterf.end(); applier++) {
+				RestoreApplierInterface &applierInterf = applier->second;
+	
+				printf("[CMD] Node:%s Notify node:%s to apply mutations to DB\n", self->describeNode().c_str(), applier->first.toString().c_str());
+				cmdReplies.push_back( applier->second.applyToDB.getReply(RestoreSimpleRequest(self->cmdID)) );
```

Comment:
I looked around, and oddly can't seem to find it already existing, but I feel like if you wrote a generic actor:

```
template <typename Interface, >
std::vector<REPLY_TYPE(T)> getBroadcastReply( std::vector<RequestStream<T>> requestStreams, const T& request ) {
  std::vector<REPLY_TYPE(T)> replies;
  // getReply all the requestStreams
  return getAll( replies );
}
```

(Or maybe one that takes a vector of interfaces, a member pointer, and a request (look at `loadBalance` for an example), as you seem to generally iterate though a vector of interfaces anyway.)

You could probably save yourself a lot of code.